File: gateway/internal/metrics/metrics.go
// Package metrics exposes JSON on /metrics.json, Prometheus text on /metrics,
// and liveness on /healthz. Stdlib http only.
package metrics

import (
	"encoding/json"
	"fmt"
	"net/http"
	"sort"
	"strings"
	"sync"
)

// sysMetrics holds per-compute-key counters.
type sysMetrics struct {
	Connections int   `json:"connections"`
	Active      int   `json:"active"`
	Wakes       int   `json:"wakes"`
	LastWakeMs  int64 `json:"last_wake_ms"`
}

// Metrics holds gateway counters, safe for concurrent use.
type Metrics struct {
	mu sync.Mutex

	ConnectionsTotal         int                    `json:"connections_total"`
	ActiveConnections        int                    `json:"active_connections"`
	WakesTotal               int                    `json:"wakes_total"`
	WakeFailuresTotal        int                    `json:"wake_failures_total"`
	SleepsTotal              int                    `json:"sleeps_total"`
	RejectedConnectionsTotal int                    `json:"rejected_connections_total"`
	WakeLatencyMsLast        int64                  `json:"wake_latency_ms_last"`
	WakeLatencyMs            []int64                `json:"wake_latency_ms"`
	PerSystem                map[string]*sysMetrics `json:"per_system"`
}

// NewMetrics constructs an empty Metrics.
func NewMetrics() *Metrics {
	return &Metrics{
		WakeLatencyMs: []int64{},
		PerSystem:     map[string]*sysMetrics{},
	}
}

// sys returns the per-system entry, creating it if needed. Caller holds the lock.
func (m *Metrics) sys(key string) *sysMetrics {
	s := m.PerSystem[key]
	if s == nil {
		s = &sysMetrics{}
		m.PerSystem[key] = s
	}
	return s
}

func (m *Metrics) ConnOpen(key string) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.ConnectionsTotal++
	m.ActiveConnections++
	s := m.sys(key)
	s.Connections++
	s.Active++
}

func (m *Metrics) ConnClose(key string) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.ActiveConnections--
	m.sys(key).Active--
}

func (m *Metrics) Wake(key string, ms int64) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.WakesTotal++
	m.WakeLatencyMsLast = ms
	m.WakeLatencyMs = append(m.WakeLatencyMs, ms)
	if len(m.WakeLatencyMs) > 100 {
		m.WakeLatencyMs = m.WakeLatencyMs[1:]
	}
	s := m.sys(key)
	s.Wakes++
	s.LastWakeMs = ms
}

func (m *Metrics) WakeFailure() {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.WakeFailuresTotal++
}

// RejectConn counts a connection refused by the GW_MAX_CONNS cap.
func (m *Metrics) RejectConn() {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.RejectedConnectionsTotal++
}

func (m *Metrics) Sleep() {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.SleepsTotal++
}

// Thread-safe accessors (used by tests and callers).
func (m *Metrics) Connections() int  { m.mu.Lock(); defer m.mu.Unlock(); return m.ConnectionsTotal }
func (m *Metrics) Active() int       { m.mu.Lock(); defer m.mu.Unlock(); return m.ActiveConnections }
func (m *Metrics) Wakes() int        { m.mu.Lock(); defer m.mu.Unlock(); return m.WakesTotal }
func (m *Metrics) WakeFailures() int { m.mu.Lock(); defer m.mu.Unlock(); return m.WakeFailuresTotal }
func (m *Metrics) Sleeps() int       { m.mu.Lock(); defer m.mu.Unlock(); return m.SleepsTotal }
func (m *Metrics) Rejected() int     { m.mu.Lock(); defer m.mu.Unlock(); return m.RejectedConnectionsTotal }

// PromText renders the Prometheus text exposition.
func (m *Metrics) PromText() string {
	m.mu.Lock()
	defer m.mu.Unlock()
	lines := []string{
		fmt.Sprintf("pggw_connections_total %d", m.ConnectionsTotal),
		fmt.Sprintf("pggw_active_connections %d", m.ActiveConnections),
		fmt.Sprintf("pggw_wakes_total %d", m.WakesTotal),
		fmt.Sprintf("pggw_wake_failures_total %d", m.WakeFailuresTotal),
		fmt.Sprintf("pggw_sleeps_total %d", m.SleepsTotal),
		fmt.Sprintf("pggw_rejected_connections_total %d", m.RejectedConnectionsTotal),
		fmt.Sprintf("pggw_wake_latency_ms_last %d", m.WakeLatencyMsLast),
	}
	keys := make([]string, 0, len(m.PerSystem))
	for k := range m.PerSystem {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	for _, k := range keys {
		s := m.PerSystem[k]
		lines = append(lines,
			fmt.Sprintf("pggw_system_active_connections{system=%q} %d", k, s.Active),
			fmt.Sprintf("pggw_system_wakes_total{system=%q} %d", k, s.Wakes),
			fmt.Sprintf("pggw_system_last_wake_ms{system=%q} %d", k, s.LastWakeMs),
		)
	}
	return strings.Join(lines, "\n") + "\n"
}

// Handler serves /healthz, /metrics.json and /metrics.
func (m *Metrics) Handler() http.Handler {
	mux := http.NewServeMux()
	mux.HandleFunc("/healthz", func(w http.ResponseWriter, _ *http.Request) {
		w.WriteHeader(http.StatusOK)
		_, _ = w.Write([]byte("ok\n"))
	})
	mux.HandleFunc("/metrics.json", func(w http.ResponseWriter, _ *http.Request) {
		m.mu.Lock()
		b, err := json.MarshalIndent(m, "", "  ")
		m.mu.Unlock()
		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		w.Header().Set("content-type", "application/json")
		w.WriteHeader(http.StatusOK)
		_, _ = w.Write(b)
	})
	mux.HandleFunc("/metrics", func(w http.ResponseWriter, _ *http.Request) {
		w.Header().Set("content-type", "text/plain")
		w.WriteHeader(http.StatusOK)
		_, _ = w.Write([]byte(m.PromText()))
	})
	return mux
}

File: gateway/src/metrics.js
'use strict';
// Minimal metrics: JSON on /metrics.json, Prometheus text on /metrics,
// liveness on /healthz. Stdlib http only.

const http = require('http');

function makeMetrics() {
  const m = {
    connections_total: 0,
    active_connections: 0,
    wakes_total: 0,
    wake_failures_total: 0,
    sleeps_total: 0,
    wake_latency_ms_last: 0,
    wake_latency_ms: [], // last 100 samples
    per_system: {},      // key -> {connections, active, wakes, last_wake_ms}
  };

  const sys = (key) => (m.per_system[key] ??= { connections: 0, active: 0, wakes: 0, last_wake_ms: 0 });

  return {
    data: m,
    connOpen(key) { m.connections_total++; m.active_connections++; const s = sys(key); s.connections++; s.active++; },
    connClose(key) { m.active_connections--; sys(key).active--; },
    wake(key, ms) {
      m.wakes_total++; m.wake_latency_ms_last = ms;
      m.wake_latency_ms.push(ms);
      if (m.wake_latency_ms.length > 100) m.wake_latency_ms.shift();
      const s = sys(key); s.wakes++; s.last_wake_ms = ms;
    },
    wakeFailure() { m.wake_failures_total++; },
    sleep() { m.sleeps_total++; },
  };
}

function promText(m) {
  const lines = [
    `pggw_connections_total ${m.connections_total}`,
    `pggw_active_connections ${m.active_connections}`,
    `pggw_wakes_total ${m.wakes_total}`,
    `pggw_wake_failures_total ${m.wake_failures_total}`,
    `pggw_sleeps_total ${m.sleeps_total}`,
    `pggw_wake_latency_ms_last ${m.wake_latency_ms_last}`,
  ];
  for (const [k, s] of Object.entries(m.per_system)) {
    lines.push(`pggw_system_active_connections{system="${k}"} ${s.active}`);
    lines.push(`pggw_system_wakes_total{system="${k}"} ${s.wakes}`);
    lines.push(`pggw_system_last_wake_ms{system="${k}"} ${s.last_wake_ms}`);
  }
  return lines.join('\n') + '\n';
}

function startMetricsServer(metrics, port, host = '0.0.0.0') {
  const server = http.createServer((req, res) => {
    if (req.url === '/healthz') { res.writeHead(200); res.end('ok\n'); return; }
    if (req.url === '/metrics.json') {
      res.writeHead(200, { 'content-type': 'application/json' });
      res.end(JSON.stringify(metrics.data, null, 2));
      return;
    }
    if (req.url === '/metrics') {
      res.writeHead(200, { 'content-type': 'text/plain' });
      res.end(promText(metrics.data));
      return;
    }
    res.writeHead(404); res.end();
  });
  server.listen(port, host);
  return server;
}

module.exports = { makeMetrics, startMetricsServer, promText };
